cmd/kubeadm/app/cmd/join.go

```diff
@@ -413,87 +412,10 @@ func (j *joinData) OutputWriter() io.Writer {
 	return j.outputWriter
 }
 
-// Run executes worker node provisioning and tries to join an existing cluster.
-func (j *joinData) Run() error {
-	// Fetch the init configuration based on the join configuration.
-	// TODO: individual phases should call these:
-	//   - phases that need initCfg should call joinData.InitCfg().
-	//   - phases that need tlsBootstrapCfg should call joinData.TLSBootstrapCfg().
-	initCfg, err := j.InitCfg()
-	if err != nil {
-		return err
-	}
-
-	// if the node is hosting a new control plane instance
-	if j.cfg.ControlPlane != nil {
-		// Completes the control plane setup
-		if err := j.PostInstallControlPlane(initCfg); err != nil {
-			return err
-		}
-
-		// outputs the join control plane done template and exits
-		etcdMessage := ""
-		// in case of local etcd
-		if initCfg.Etcd.External == nil {
-			etcdMessage = "* A new etcd member was added to the local/stacked etcd cluster."
-		}
-
-		ctx := map[string]string{
-			"KubeConfigPath": kubeadmconstants.GetAdminKubeConfigPath(),
-			"etcdMessage":    etcdMessage,
-		}
-		joinControPlaneDoneTemp.Execute(j.outputWriter, ctx)
-		return nil
-	}
-
-	// otherwise, if the node joined as a worker node;
-	// outputs the join done message and exits
-	fmt.Fprintf(j.outputWriter, joinWorkerNodeDoneMsg)
```

Comment:
If I'm not wrong, this code is not finding a new home after the proposed refactor.
If this is true, please implement a solution similar to the one implemented in init (where the Run method of the cobra command triggers the print of the final message)

Reply:
Yes, this should be moved to the `join` toplevel `Run` function, and only print it if this is not a master node (as the removed logic implemented).

Reply:
done! similar to one implemented in init